### PR TITLE
header support implementation

### PR DESCRIPTION
**Pull request by accelerated**
Hi @mfontanini, please take a look. This is just the first pass of the implementation for review purposes only. I haven't written any tests yet and also have not decorated methods with doxygen comments as things may change. 

Some considerations:
1) I have not implemented the name-based getter function `rd_kafka_header_get` the reason being that the more generic `rd_kafka_header_get_all` can be used as a forward/reverse iterator and can actually have an `end()` function which can be used for bound checking when iterating and matches well with STL container API. With the name based getter, there is no way of guaranteeing what `end()` will be, hence iterating can only be implemented as a forward iterator with a while loop. I felt that in any case, the number of headers won't be very large and anyone using the generic iterator over all elements can simply check for the header name they want, so there's no loss in functionality and the code is less complicated.
The _rdkafka_ iterator API is strange because it lets you iterate by index, but there's no API to delete a header at a certain position, instead it deletes by name, which could have duplicates, in which case one delete operation can remove entire portions of the underlying vector (I checked the _rdkafka_ code and the header list is indeed a vector).
2) `HeaderList<HeaderType> Message::detach_header_list()` function could be implemented in a more c++ way as `HeaderList<HeaderType> Message::get_header_list() &&` which would be an overload of the const `get_header_list()` but would apply only to rvalue messages. The effect would be the same. Not sure what your preference is. 
3) The is a potential performance issue with the `Producer::produce(const MessageBuilder)` and `Producer::produce(const Message)` functions. The problem is that `rd_kafka_producev()` takes ownership of the header list and because both `MessageBuilder` and `Message` are const, you can't make the underlying handle _non-owning_ (which would be the fastest). Also a user might produce multiple messages with the same builder object so you have to maintain ownership in the `const` case,
which forces you to clone the header list each time. Perhaps we can add two more produce functions with move semantics which would take `MessageBuilder&&` and `Message&&` ? 
4) Inside the `HeaderList` class, I made the following assumptions. If the list object is _non-owning_, the copy would _not_ invoke clone. If on the other hand you _own_ the handle, then a copy operation would clone the headers. You can also have an _owning_ object which on copy simply returns a _non-owning_ object, but I don't think it's the proper logic.
5) I have also added the `Message::get_latency()` api which was missing.
6) I will split the HeaderIterator and HeaderList into separate files in the next commit. 
7) Note that `Message` has a `HeaderList<Buffer>` type which is non-owning, whereas `MessageBuilder` has a `HeaderList<T>` where **T** matches the `BufferType` template parameter.

### CODE REVIEW

**Comment by mfontanini** [comment]
Sweet! Thanks for the PR, this is great. I'll have a look at it in more detail later but on a first pass it seems alright.

**Comment by mfontanini** [comment]
I presume given point 1, there can be multiple headers with the same name?

**Comment by accelerated** [comment]
Yes, any number. The rdkafka list is a vector. Any add or remove operations will invalidate iterators.

File: include/cppkafka/message.h [hunk 1]
```diff
@@ -84,52 +87,69 @@ class CPPKAFKA_API Message {
...
+     */
+    const HeaderListType& get_header_list() const {
+        return header_list_;
+    }
+    
+    /**
+     * \brief Detaches the message's header list
+     */
+    template <typename HeaderType>
+    HeaderList<HeaderType> detach_header_list() {
```
Comment (mfontanini) [inline include/cppkafka/message.h]: Following up on your comment, I think this is fine. I'd rather be explicit about wanting to detach them than having to do `move(message).detach_header_list()`.

File: include/cppkafka/header.h [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+bool operator>=(const Header<BufferType>& lhs, const Header<BufferType>& rhs) {
+    return !(lhs < rhs);
+}
+
+template <typename BufferType>
+Header<BufferType>::Header(const std::string& name,
+                           const BufferType& value)
+: name_(name),
+  value_(make_value(value)) {
+  assert(!name.empty());
```
[the remark this answers is not in the document]
Comment (accelerated) [reply include/cppkafka/header.h]: Ok I will remove the assert and I will add a null-name test to see if kafka likes it.

File: include/cppkafka/message.h [hunk 3]
```diff
@@ -152,23 +172,31 @@ class CPPKAFKA_API Message {
      * If calling rd_kafka_message_timestamp returns -1, then boost::none_t will be returned.
      */
     inline boost::optional<MessageTimestamp> get_timestamp() const;
+    
+    /**
+     * \brief Gets the message latency in microseconds as measured from the produce() call.
+     */
+    int64_t get_latency() const {
```
[the remark this answers is not in the document]
Comment (accelerated) [reply include/cppkafka/message.h]: ok

File: include/cppkafka/header_list.h [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+    {
+        HeaderIterator tmp(*this);
+        operator++();
+        return tmp;
+    }
+    
+    // prefix decrement
+    HeaderIterator& operator--()
+    {
+        if (index_ == 0) {
```
[the remark this answers is not in the document]
Comment (accelerated) [reply include/cppkafka/header_list.h]: ok

File: include/cppkafka/header_list.h [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+    HeaderList();
+    explicit HeaderList(size_t reserve);
+    explicit HeaderList(const rd_kafka_headers_t* handle);
+    
+    Error add(const HeaderType& header);
+    Error remove(const std::string& name);
+    HeaderType at(size_t index) const; //throws
+    HeaderType front() const;
+    HeaderType back() const;
+    HeaderType back(const std::string& name) const;
```
[the remark this answers is not in the document]
Comment (accelerated) [reply include/cppkafka/header_list.h]: yeah i just had it for completeness reasons, but I agree with you, seems like a totally useless function. I'll remove it.

File: include/cppkafka/header.h [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+namespace cppkafka {
+
+template <typename BufferType>
+class Header {
+public:
+    using ValueType = BufferType;
+    Header() = default;
+    
+    Header(const std::string& name,
```
[the remark this answers is not in the document]
Comment (accelerated) [reply include/cppkafka/header.h]: BufferType is also stored...shouldn't in this case be passed by value and moved ?

**Comment by accelerated** [comment]
Regarding ... 
>Alright, I managed to get back to this and have a look at the code. It looks good, just a few comments. Regarding your third comment about the produce performance, I think that's a good idea. Maybe having something like this would work?
```
void produce(MessageBuilder&& b) {
    // move them
    do_produce(b, move(b.header_list());
}

void produce(const MessageBuilder& b) {
    // copy them
    do_produce(MessageBuilder::HeaderListType(b.header_list));
}
```
no this will not work. What needs to be done for the `MessageBuilder` class is to have a method to `detach` the header list handle, and then create a non-owning header list with that which you then pass into `producev`. Simply moving the list will not work, because moving transfers ownership of the `unique_ptr` elsewhere...what we need to do is get rid of that ownership altogether since rdkafka will assume it. So the `HeaderList` class must have a method `rdkafka_header_list_t* HeaderList::detach()` which calls `release()` on the pointer and returns the handle and then you can do : `make_non_owning(b.header_list().detach())`.

In the case of `Message&&`, since `Message` object is non-owning, you can't just pass the handle into rdkafka as it will take ownership and now the handle will have 2 owners. In this case, you need to detach the list (like above), and pass it into rdkafka. Ex: `make_non_owning(message.detach_header_list().detach())` which calls the same `HeaderList::detach()` method. One thing to remember is that `Message::detach_header_list()` creates an owning header list so that in turn must also be detached. 

For the const ref methods you have to clone. There's not option there.

**Comment by mfontanini** [comment]
I see your point, makes sense. I think I would go the `release_ownership` way but I'd call it `release` so it's in line with `unique_ptr`/`shared_ptr`.

File: include/cppkafka/header.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (accelerated) [reply include/cppkafka/header.h]: Can you pls comment on this...I'm not sure if you also want the BufferType to be passed by value instead of lvalue/rvalue ?

[the remark this answers is not in the document]
Comment (mfontanini) [reply include/cppkafka/header.h]: Given a `BufferType` can't be moved automatically, I don't think this one should take a value. Probably just have both constructors take a string by value.

**Comment by mfontanini** [comment]
Hmm it's weird because the last two ([#1](https://travis-ci.org/mfontanini/cppkafka/jobs/439761871) and [#2](https://travis-ci.org/mfontanini/cppkafka/jobs/439768703)) builds failed on exactly the same tests. If this was a timing issue I'd expect it to be more unpredictable. 

Unrelated to failing tests but I forgot to mention: the header support should be allowed to be disabled via some compilation flag. Otherwise nobody would be able to use this with rdkafka < whatever version header support was introduced in.

**Comment by accelerated** [comment]
> Hmm it's weird because the last two ([#1](https://travis-ci.org/mfontanini/cppkafka/jobs/439761871) and [#2](https://travis-ci.org/mfontanini/cppkafka/jobs/439768703)) builds failed on exactly the same tests. If this was a timing issue I'd expect it to be more unpredictable.
> 
> Unrelated to failing tests but I forgot to mention: the header support should be allowed to be disabled via some compilation flag. Otherwise nobody would be able to use this with rdkafka < whatever version header support was introduced in.

Yeah makes sense...we can do something similar to the Queue memory leak issue. I'll fix it. The producer code will look somewhat uglier as that's the entry point of use...

**Comment by accelerated** [comment]
Good to go!

**Comment by accelerated** [comment]
It might well be that the rdkafka changes from 0.11.0 to 0.11.5 make a difference in timing. They made many perf improvements, hence why the same tests failed. Your consumer runner class might need a tweak perhaps?

File: src/producer.cpp [hunk 7]
```diff
@@ -65,6 +65,42 @@ Producer::PayloadPolicy Producer::get_payload_policy() const {
...
+    do_produce(temp, MessageBuilder::HeaderListType(temp.header_list().release_handle())); //move headers
+}
+
+void Producer::produce(const Message& message) {
+    do_produce(message, HeaderList<Message::HeaderType>(message.get_header_list())); //copy headers
+}
+
+void Producer::produce(Message&& message) {
+    Message temp(std::move(message)); //rdakfka still owns the header list at this point
+    do_produce(temp, HeaderList<Message::HeaderType>(temp.detach_header_list<Message::HeaderType>())); //move headers
```
Comment (mfontanini) [inline src/producer.cpp]: I think the `HeaderList<Message::HeaderType>` is unnecessary, unless I'm missing something.

File: tests/producer_test.cpp [hunk 8]
```diff
@@ -315,6 +350,50 @@ TEST_CASE("multiple messages", "[producer]") {
...
+    REQUIRE(message.get_header_list().size() == 3);
+    CHECK(message.get_header_list().front() == Hdr());
+    CHECK(message.get_header_list().at(1) == Hdr(std::string(""), header2));
+    CHECK(message.get_header_list().back() == Hdr(std::string("header3"), header3));
+    
+    //validate second message
+    CHECK(messages[0].get_header_list() == messages[1].get_header_list());
+    CHECK(messages[0].get_header_list().get_handle() != messages[1].get_header_list().get_handle());
+}
+
```
Comment (mfontanini) [inline tests/producer_test.cpp]: Good job on the tests!

File: src/producer.cpp [hunk 9]
```diff
@@ -65,6 +65,42 @@ Producer::PayloadPolicy Producer::get_payload_policy() const {
 }
 
 void Producer::produce(const MessageBuilder& builder) {
+    do_produce(builder, MessageBuilder::HeaderListType(builder.header_list())); //copy headers
+}
+
+void Producer::produce(MessageBuilder&& builder) {
+    MessageBuilder temp(std::move(builder)); //owns header list after the move
+    do_produce(temp, MessageBuilder::HeaderListType(temp.header_list().release_handle())); //move headers
```
[the remark this answers is not in the document]
Comment (accelerated) [reply src/producer.cpp]: Yeah that was the reason, I wanted to make sure a moved object will be empty afterwards (not just the headers) regardless if it's called from a temp rvalue or from a moved lvalue. Ok will remove it.

File: include/cppkafka/header.h [hunk 10]
```diff
@@ -36,24 +36,60 @@
...
+    Header(const std::string name,
            const BufferType& value);
     
-    Header(const std::string& name,
+    /**
+     * \brief Build a header instance
+     * \param name The header name
+     * \param value The header data to be moved
+     */
+    Header(const std::string name,
```
[the remark this answers is not in the document]
Comment (accelerated) [reply include/cppkafka/header.h]: Silly me. Of course...good catch!

File: src/producer.cpp (hunk 7 above)
Comment (accelerated) [reply src/producer.cpp]: You do need it, for `Message::get_header_list()` and `Message::detach_header_list()` are templetized on the type of header you want. `Message` itself has `Header<Buffer>` internally but there's no reason you can request a list as `Header<std::string>` out of it. For example, you can get the list as `Header<std::string>`, modify it and create a new builder from it.

Comment (accelerated) [reply src/producer.cpp]: If you think that's too much overkill, I can remove the template and always force the user to get a `HeaderList<Header<Buffer>>` in which case if they want to modify it, they will need to create their own:
```
HeaderList<Header<std::string>> list(make_non_owning(message.detach_header_list().get_handle()));
```

Comment (accelerated) [reply src/producer.cpp]: Yes you're right, I'll remove it.

**Comment by mfontanini** [comment]
I think this looks good! I'm going to clone it over the weekend, play around a bit and see if I can fix the test issues and we can probably merge it after that.

File: tests/headers_test.cpp [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+        CHECK(list.at(1) == list2.at(1));
+        CHECK(list.at(2) == list2.at(2));
+        CHECK(list == list2);
+    }
+    
+    SECTION("copy owning with buffers") {
+        //Create an owning header list and copy it
+        HeaderList<BufferHeader> list(3), list2(3);
+        string payload1 = "payload1", payload2 = "payload2", payload3 = "payload3";
+        list.add(BufferHeader(std::string("header1"), payload1));
```
[the remark this answers is not in the document]
Comment (accelerated) [reply tests/headers_test.cpp]: Sure but how will the compiler know the instantiate a `BufferHeader` vs a `StringHeader` with that initializer list? Both can be created with the same arguments.

[the remark this answers is not in the document]
Comment (mfontanini) [reply tests/headers_test.cpp]: `HeaderList::add` takes a `HeaderType` which for that particular instantiation (`list`), that type is `BufferHeader` so there's no ambiguity of which one you're referring to.

**Comment by accelerated** [comment]
Instead of using a consumer runner, can't you produce say N messages, then start the consumer with relative offsets (last-N) since there are no local offsets cached at this point, read N messages till EOF and then compare? Then you eliminate the runner altogether.

**Comment by mfontanini** [comment]
Well, some tests require simply subscribing and there could be some uncommitted messages in the topic you subscribe to. Also you need to make sure that you get the messages at some point, as if there's some bug and you don't produce them, you don't want the runner to be stuck forever. It's a pretty hacky solution but it kinda worked until it didn't.

**Comment by accelerated** [comment]
The consumer block can be solved easily by only polling x amount of times or with a max timer. I'm not sure what happens if you have uncomitted messages in the topic, and you open the consumer with relative offset "latest". Is it latest committed + 1 or just latest timewise, aka top of the queue? 
Could be worth testing. Alternatively, you create 500 topics and each test uses a fresh topic. Clean slate.

**Comment by mfontanini** [comment]
Creating 500 topics will still be an issue if a previous run on the same topic fails. Plus, that's not very maintainable :P. If you keep using the same consumer group id, the auto offset reset policy is irrelevant because that's a fallback in case the consumer doesn't have any already committed offset for a topic/partition. Since there's been previous runs of the tests and offsets were committed, this is irrelevant. Now, if you used a random consumer group id on each run and use reset policy "latest", that would work. I'm not a fan of this approach but if it's the only way to make it work reliably then that's fine.

**Comment by accelerated** [comment]
Yeah i see what you mean. Ideally a consumer group id made from a GUID would be unique for each test, esp a v1Guid which is time based. Alternatively maybe blowing away the topics and recreating them  on startup ensures everything is empty. I was assuming the travis sandbox would be a blank state each time you run kafka, meaning no topic or partitions leftovers from previous runs.

**Comment by mfontanini** [comment]
Travis does run in a clean state, but the issue is that all tests share the same topics and consumer groups. I think using a random string as the consumer group is enough here. Something like 16 alphanums generated via `random_device` + `mt19937` + `uniform_int_distribution` should be enough.

**Comment by accelerated** [comment]
Right, each test gets a new producer, consumer pair. One other possibility is to implement the admin api and generate/teardown topics dynamically? Not too familiar with this api though...
In any case, making this test rehaul is a big operation, i wonder if you can merge the allocator and headers PRs and have the test refactor as a new issue where you can ask for help? For now my bandwidth is very limited....

**Comment by mfontanini** [comment]
Yeah, I think we can merge this. I wouldn't go the admin API way because that only works in some recent version of rdkafka so you wouldn't be able to test it locally. Can you just rebase this one? I should have some time in the afternoon today or tomorrow so I can give it a try.

**Comment by mfontanini** [comment]
Thanks!
